File: archive/pulumi-ts/Pr7298/lib/waf-stack.ts
/**
 * waf-stack.ts
 *
 * Defines AWS WAFv2 Web ACL for protecting Application Load Balancer.
 *
 * Features:
 * - Rate limiting: 1000 requests per 5 minutes per IP
 * - AWS Managed Rule Groups (Common Rule Set, Known Bad Inputs)
 * - Optional geo-blocking
 * - CloudWatch metrics for all rules
 */
import * as pulumi from '@pulumi/pulumi';
import * as aws from '@pulumi/aws';

export interface WafStackArgs {
  environmentSuffix: string;
  albArn?: pulumi.Input<string>; // ALB ARN to associate WAF
  enableGeoBlocking?: boolean; // Optional geo-blocking
  blockedCountries?: string[]; // ISO country codes to block
  tags?: pulumi.Input<{ [key: string]: string }>;
}

export class WafStack extends pulumi.ComponentResource {
  public readonly webAcl: aws.wafv2.WebAcl;
  public readonly webAclArn: pulumi.Output<string>;
  public readonly webAclAssociation?: aws.wafv2.WebAclAssociation;

  constructor(
    name: string,
    args: WafStackArgs,
    opts?: pulumi.ComponentResourceOptions
  ) {
    super('tap:waf:WafStack', name, args, opts);

    const {
      environmentSuffix,
      albArn,
      enableGeoBlocking = false,
      blockedCountries = [],
      tags,
    } = args;

    // Define WAF rules
    const rules: aws.types.input.wafv2.WebAclRule[] = [
      // Rule 0: Rate-based rule (1000 requests per 5 minutes)
      {
        name: 'RateLimitRule',
        priority: 0,
        statement: {
          rateBasedStatement: {
            limit: 1000,
            aggregateKeyType: 'IP',
          },
        },
        action: {
          block: {},
        },
        visibilityConfig: {
          sampledRequestsEnabled: true,
          cloudwatchMetricsEnabled: true,
          metricName: 'RateLimitRule',
        },
      },
      // Rule 1: AWS Managed Rules - Common Rule Set
      {
        name: 'AWSManagedRulesCommon',
        priority: 1,
        statement: {
          managedRuleGroupStatement: {
            vendorName: 'AWS',
            name: 'AWSManagedRulesCommonRuleSet',
          },
        },
        overrideAction: {
          none: {},
        },
        visibilityConfig: {
          sampledRequestsEnabled: true,
          cloudwatchMetricsEnabled: true,
          metricName: 'AWSManagedRulesCommon',
        },
      },
      // Rule 2: AWS Managed Rules - Known Bad Inputs
      {
        name: 'AWSManagedRulesKnownBadInputs',
        priority: 2,
        statement: {
          managedRuleGroupStatement: {
            vendorName: 'AWS',
            name: 'AWSManagedRulesKnownBadInputsRuleSet',
          },
        },
        overrideAction: {
          none: {},
        },
        visibilityConfig: {
          sampledRequestsEnabled: true,
          cloudwatchMetricsEnabled: true,
          metricName: 'AWSManagedRulesKnownBadInputs',
        },
      },
    ];

    // Add geo-blocking rule if enabled
    if (enableGeoBlocking && blockedCountries.length > 0) {
      rules.push({
        name: 'GeoBlockingRule',
        priority: 3,
        statement: {
          geoMatchStatement: {
            countryCodes: blockedCountries,
          },
        },
        action: {
          block: {},
        },
        visibilityConfig: {
          sampledRequestsEnabled: true,
          cloudwatchMetricsEnabled: true,
          metricName: 'GeoBlockingRule',
        },
      });
    }

    // Create WAFv2 Web ACL
    this.webAcl = new aws.wafv2.WebAcl(
      `cicd-waf-${environmentSuffix}`,
      {
        name: `cicd-waf-${environmentSuffix}`,
        scope: 'REGIONAL', // For ALB/API Gateway (use CLOUDFRONT for CloudFront)
        defaultAction: {
          allow: {}, // Default allow, rules will block bad traffic
        },
        rules,
        visibilityConfig: {
          sampledRequestsEnabled: true,
          cloudwatchMetricsEnabled: true,
          metricName: `cicd-waf-${environmentSuffix}`,
        },
        tags: {
          ...tags,
          Name: `cicd-waf-${environmentSuffix}`,
        },
      },
      { parent: this }
    );

    this.webAclArn = this.webAcl.arn;

    // Associate WAF with ALB if provided
    if (albArn) {
      this.webAclAssociation = new aws.wafv2.WebAclAssociation(
        `alb-waf-association-${environmentSuffix}`,
        {
          resourceArn: albArn,
          webAclArn: this.webAcl.arn,
        },
        { parent: this }
      );
    }

    this.registerOutputs({
      webAclId: this.webAcl.id,
      webAclArn: this.webAcl.arn,
      webAclName: this.webAcl.name,
    });
  }
}

File: archive/pulumi-py/Pr3797/lib/waf_stack.py
"""
WAF Stack for CloudFront security and geo-blocking.
"""

import pulumi
from pulumi_aws import wafv2, Provider
from pulumi import ResourceOptions
from typing import Optional


class WAFStack(pulumi.ComponentResource):
    """
    Creates AWS WAF Web ACL with geo-blocking rules for CloudFront.
    Note: WAF for CloudFront must be created in us-east-1 region.
    """

    def __init__(
        self,
        name: str,
        environment_suffix: str,
        tags: dict,
        opts: Optional[ResourceOptions] = None
    ):
        super().__init__('tap:waf:WAFStack', name, None, opts)

        # Create provider for us-east-1 (required for CloudFront WAF)
        us_east_1_provider = Provider(
            f"waf-us-east-1-provider-{environment_suffix}",
            region="us-east-1",
            opts=ResourceOptions(parent=self)
        )

        # Create IP Set for additional blocking if needed
        ip_set = wafv2.IpSet(
            f"waf-ip-set-{environment_suffix}",
            name=f"tap-blocked-ips-{environment_suffix}",
            scope="CLOUDFRONT",
            ip_address_version="IPV4",
            addresses=[],
            tags=tags,
            opts=ResourceOptions(parent=self, provider=us_east_1_provider)
        )

        # Create Web ACL with geo-blocking
        self.web_acl = wafv2.WebAcl(
            f"cloudfront-web-acl-{environment_suffix}",
            name=f"tap-cloudfront-acl-{environment_suffix}",
            scope="CLOUDFRONT",
            default_action=wafv2.WebAclDefaultActionArgs(
                allow=wafv2.WebAclDefaultActionAllowArgs()
            ),
            rules=[
                # Geo-blocking rule
                wafv2.WebAclRuleArgs(
                    name="geo-blocking-rule",
                    priority=1,
                    action=wafv2.WebAclRuleActionArgs(
                        block=wafv2.WebAclRuleActionBlockArgs()
                    ),
                    statement=wafv2.WebAclRuleStatementArgs(
                        geo_match_statement=wafv2.WebAclRuleStatementGeoMatchStatementArgs(
                            country_codes=["CN", "RU", "KP"]
                        )
                    ),
                    visibility_config=wafv2.WebAclRuleVisibilityConfigArgs(
                        cloudwatch_metrics_enabled=True,
                        metric_name="geo-blocking-rule",
                        sampled_requests_enabled=True
                    )
                ),
                # Rate limiting rule
                wafv2.WebAclRuleArgs(
                    name="rate-limiting-rule",
                    priority=2,
                    action=wafv2.WebAclRuleActionArgs(
                        block=wafv2.WebAclRuleActionBlockArgs()
                    ),
                    statement=wafv2.WebAclRuleStatementArgs(
                        rate_based_statement=wafv2.WebAclRuleStatementRateBasedStatementArgs(
                            limit=2000,
                            aggregate_key_type="IP"
                        )
                    ),
                    visibility_config=wafv2.WebAclRuleVisibilityConfigArgs(
                        cloudwatch_metrics_enabled=True,
                        metric_name="rate-limiting-rule",
                        sampled_requests_enabled=True
                    )
                ),
                # AWS managed rule - Common Rule Set
                wafv2.WebAclRuleArgs(
                    name="aws-managed-common-rule",
                    priority=3,
                    override_action=wafv2.WebAclRuleOverrideActionArgs(
                        none=wafv2.WebAclRuleOverrideActionNoneArgs()
                    ),
                    statement=wafv2.WebAclRuleStatementArgs(
                        managed_rule_group_statement=wafv2.WebAclRuleStatementManagedRuleGroupStatementArgs(
                            name="AWSManagedRulesCommonRuleSet",
                            vendor_name="AWS"
                        )
                    ),
                    visibility_config=wafv2.WebAclRuleVisibilityConfigArgs(
                        cloudwatch_metrics_enabled=True,
                        metric_name="aws-managed-common-rule",
                        sampled_requests_enabled=True
                    )
                )
            ],
            visibility_config=wafv2.WebAclVisibilityConfigArgs(
                cloudwatch_metrics_enabled=True,
                metric_name=f"tap-cloudfront-acl-{environment_suffix}",
                sampled_requests_enabled=True
            ),
            tags=tags,
            opts=ResourceOptions(parent=self, provider=us_east_1_provider)
        )

        self.web_acl_id = self.web_acl.arn

        self.register_outputs({
            'web_acl_id': self.web_acl_id
        })
